### PR TITLE
feat: enkryptWallet

### PR DESCRIPTION
**Pull request by kvhnuke**
Issues on our previous PR #917 should be fixed on this.
However, it is worth noting that multiple connect popups were due to some kind of error mishandling on
`import { createClient} from 'wagmi'`
Basically, when `eth_accounts` call get rejected, itll try again for exactly 3 times. 

Now the behavior is, itll do `eth_accounts` if the user deny it it'll try to do `eth_requestAccounts` and if the user deny that too, itll stop. It will no longer continue to try to connect

### CODE REVIEW

**Comment by DanielSinclair** [comment]
@kvhnuke Looks like we're still getting some TypeScript/CI errors here. A `pnpm run lint:fix` should correct one of them.
```
packages/rainbowkit prepare: src/wallets/walletConnectors/enkryptWallet/enkryptWallet.ts(63,33): error TS2339: Property 'getProvider' does not exist on type 'EIP5749Connector'.
--
00:28:25.649 | packages/rainbowkit prepare: src/wallets/walletConnectors/enkryptWallet/enkryptWallet.ts(69,13): error TS7006: Parameter 'accounts' implicitly has an 'any' type.
00:28:25.650 | packages/rainbowkit prepare: src/wallets/walletConnectors/enkryptWallet/enkryptWallet.ts(70,14): error TS7006: Parameter 'err' implicitly has an 'any' type.
Expand 24 Lines
00:28:37.067 | packages/rainbowkit prepare: src/wallets/walletConnectors/enkryptWallet/enkryptWallet.ts(63,33): error TS2339: Property 'getProvider' does not exist on type 'EIP5749Connector'.
00:28:37.068 | packages/rainbowkit prepare: src/wallets/walletConnectors/enkryptWallet/enkryptWallet.ts(69,13): error TS7006: Parameter 'accounts' implicitly has an 'any' type.
00:28:37.069 | packages/rainbowkit prepare: src/wallets/walletConnectors/enkryptWallet/enkryptWallet.ts(70,14): error TS7006: Parameter 'err' implicitly has an 'any' type.
Expand 21 Lines
00:28:49.256 | Failed to compile.
00:28:49.256 |  
00:28:49.256 | ./pages/_app.tsx
00:28:49.256 | 19:3  Error: Member 'argentWallet' of the import declaration should be sorted alphabetically.  sort-imports
00:28:49.256 |  
00:28:49.256 | info  - Need to disable some ESLint rules? Learn more here: https://nextjs.org/docs/basic-features/eslint#disabling-rules
00:28:49.271 | undefined
00:28:49.272 | /vercel/path0/packages/example:
00:28:49.272 | ERR_PNPM_RECURSIVE_RUN_FIRST_FAIL  example@0.0.48 build: `next build`
00:28:49.272 | Exit status 1
00:28:49.292 | ELIFECYCLE  Command failed with exit code 1.
00:28:49.315 | Error: Command "pnpm run ci:example" exited with 1
00:28:49.688 | BUILD_UTILS_SPAWN_1: Command "pnpm run ci:example" exited with 1
```

File: site/data/docs/custom-wallet-list.mdx
Comment (DanielSinclair) [inline site/data/docs/custom-wallet-list.mdx:154]: For simplicity, we have deprecated `shimDisconnect` from the Custom Wallet List docs

File: packages/rainbowkit/src/wallets/walletConnectors/enkryptWallet/enkryptWallet.ts
Comment (DanielSinclair) [inline packages/rainbowkit/src/wallets/walletConnectors/enkryptWallet/enkryptWallet.ts:85]: Should pass `options` directly or spread; Wagmi's `shimDisconnect` defaults will be overridden if `shimDisconnect` is undefined

File: packages/rainbowkit/src/wallets/walletConnectors/enkryptWallet/enkryptWallet.ts
Comment (DanielSinclair) [inline packages/rainbowkit/src/wallets/walletConnectors/enkryptWallet/enkryptWallet.ts:77]: Reference other connectors for how we accept `InjectedConnectorOptions` and pass through user provided options

File: packages/rainbowkit/src/wallets/walletConnectors/enkryptWallet/enkryptWallet.ts
Comment (DanielSinclair) [inline packages/rainbowkit/src/wallets/walletConnectors/enkryptWallet/enkryptWallet.ts:96]: `installed` is used incorrectly; RainbowKit uses this value to determine when to show the user the connect state versus the "Get Wallet" state. Should check if wallet is injected and provide a boolean or undefined.

File: packages/rainbowkit/src/wallets/walletConnectors/enkryptWallet/enkryptWallet.ts
Comment (DanielSinclair) [inline packages/rainbowkit/src/wallets/walletConnectors/enkryptWallet/enkryptWallet.ts:71]: I'd like to have this `eth_accounts` issue addressed in the wallet or Wagmi rather than in this connector. Haven't bumped into another wallet with this issue.

File: packages/rainbowkit/src/wallets/walletConnectors/enkryptWallet/enkryptWallet.ts
Comment (DanielSinclair) [inline packages/rainbowkit/src/wallets/walletConnectors/enkryptWallet/enkryptWallet.ts:40]: Does Enkrypt currently support `window.ethereum`, an alternate `window.wallet`, or `window.ethereum.providers`? Would like to see 5749 settled and introduced in Wagmi before we adopt, and would prefer to adopt the current standards in this connector. We will migrate all connectors to 5749 once that has moved forward.

**Review by DanielSinclair — CHANGES_REQUESTED** [review]
Looks like Enkrypt also needs a provider name introduced here in Wagmi. RainbowKit relies on a check against this name for `window.ethereum.providers` support.
https://github.com/wagmi-dev/references/blob/main/packages/connectors/src/utils/getInjectedName.ts

**Comment by DanielSinclair** [comment]
Please also allow maintainer edits for a faster merge after these issues are addressed.

File: packages/rainbowkit/src/wallets/walletConnectors/enkryptWallet/enkryptWallet.ts
Comment (kvhnuke) [reply packages/rainbowkit/src/wallets/walletConnectors/enkryptWallet/enkryptWallet.ts:96]: this is required due to the fact that, with Manifest V3 injected wallets currently it wont inject before the page load, so when the library check whether an injected wallet exists it'll return null. Thats why you have to check it after a few millisecond delay and set installed accordingly

File: packages/rainbowkit/src/wallets/walletConnectors/enkryptWallet/enkryptWallet.ts
Comment (kvhnuke) [reply packages/rainbowkit/src/wallets/walletConnectors/enkryptWallet/enkryptWallet.ts:40]: it does, however is there any specific issue regarding this approach?

**Comment by kvhnuke** [comment]
> Please also allow maintainer edits for a faster merge after these issues are addressed.

unfortunately I dont think this is possible since the fork is from an organization

File: packages/rainbowkit/src/wallets/walletConnectors/enkryptWallet/enkryptWallet.ts
Comment (DanielSinclair) [reply packages/rainbowkit/src/wallets/walletConnectors/enkryptWallet/enkryptWallet.ts:40]: @kvhnuke Only concerned about fragility. RainbowKit inherits `window.ethereum.providers` support from Wagmi. Because Enkrypt is currently unique, we would like to adopt the connector through `@wagmi/connectors` akin to the `MetaMaskConnector` or `InjectedConnector` implementations there. We can then inherit that upstream in RainbowKit for the detection logic.

File: packages/rainbowkit/src/wallets/walletConnectors/enkryptWallet/enkryptWallet.ts
Comment (DanielSinclair) [reply packages/rainbowkit/src/wallets/walletConnectors/enkryptWallet/enkryptWallet.ts:96]: @kvhnuke RainbowKit uses `installed` for UI gating, including displaying a fallback injected wallet for unsupported wallets. We do actually need that property to be accurate. Every Manifest v3 wallet that bumped into this issue did find a workaround: https://github.com/rainbow-me/rainbowkit/pull/1042#issuecomment-1505331023

Comment (kvhnuke) [reply packages/rainbowkit/src/wallets/walletConnectors/enkryptWallet/enkryptWallet.ts:96]: unfortunately those are workarounds and there is chrome bug conversation happening
finally they made the fix on Chrome 111 but current chrome version is 112, so We are waiting for atleast Chrome version 116 to release the fix
https://bugs.chromium.org/p/chromium/issues/detail?id=634381

Comment (kvhnuke) [reply packages/rainbowkit/src/wallets/walletConnectors/enkryptWallet/enkryptWallet.ts:96]: anyway Ill push a simplified code that doesnt touch installed

File: packages/rainbowkit/src/wallets/walletConnectors/enkryptWallet/enkryptWallet.ts
Comment (kvhnuke) [reply packages/rainbowkit/src/wallets/walletConnectors/enkryptWallet/enkryptWallet.ts:40]: but I see Phantom is using `window.phantom.ethereum`? will it be ok to use `window.enkrypt.providers.ethereum`?

File: packages/rainbowkit/src/wallets/walletConnectors/enkryptWallet/enkryptWallet.ts
Comment (kvhnuke) [reply packages/rainbowkit/src/wallets/walletConnectors/enkryptWallet/enkryptWallet.ts:96]: > @kvhnuke RainbowKit uses `installed` for UI gating, including displaying a fallback injected wallet for unsupported wallets. We do actually need that property to be accurate. Every Manifest v3 wallet that bumped into this issue did find a workaround: [#1042 (comment)](https://github.com/rainbow-me/rainbowkit/pull/1042#issuecomment-1505331023)

also xdefi wallet is Manifest v2

Comment (punithbm) [reply packages/rainbowkit/src/wallets/walletConnectors/enkryptWallet/enkryptWallet.ts]: Going through similar issue in the manifest v3  -  #1041

**Comment by DanielSinclair** [comment]
Closing for now until we can get upstream support into Wagmi for this type of connector (or when 6963 settles). Enkrypt is definitely one of the wallets we'd like to support with RainbowKit.

**Pull request closed without merging**
